lazy_theta_star_planner/src/lazy_theta_star2.cpp

```diff
@@ -0,0 +1,257 @@
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+#include "lazy_theta_star_planner/lazy_theta_star2.h"
+#include <vector>
+
+namespace lazyThetaStar
+{
+
+LazyThetaStar::LazyThetaStar()
+{
+  sizeX = 0;
+  sizeY = 0;
+}
+
+bool LazyThetaStar::getPath(std::vector<coords<world_pts>> & raw_path)
+{
+  RCLCPP_INFO(node_->get_logger(), "Path Planning Begins....");
+
+  initializeStuff();
+
+  std::cout << src.x << '\t' << src.y << '\t' << dst.x << '\t' << dst.y << '\n';
+
+  data.push_back({src.x, src.y, 0, dist(src.x, src.y, dst.x, dst.y),
+      0, 1, dist(src.x, src.y, dst.x, dst.y)});
+  addIndex(src.x, src.y, id_gen);
+  pq.push({id_gen, &(data[id_gen].f)});
+
+  id_gen++;
+  id curr_id = 0;
+
+  if (!losCheck(src.x, src.y, dst.x, dst.y)) {
+    while (!pq.empty()) {
+
+      tree_node curr_data = data[curr_id];
+      curr_data.closed = -1;
+      id curr_par = curr_data.parent_id;
+
+      if (isGoal(curr_data.x, curr_data.y)) {
+        if (!(losCheck(curr_data.x, curr_data.y, data[curr_par].x, data[curr_par].y))) {
+          resetParent(curr_data);
+        }
+        break;
+      }
+
+      if (!(losCheck(curr_data.x, curr_data.y, data[curr_par].x, data[curr_par].y))) {
+        resetParent(curr_data);
+      }
+
+      setNeighbors(curr_data);
+
+      curr_id = pq.top().pos_id;
+      pq.pop();
+    }
+    if (pq.empty()) {
+      RCLCPP_INFO(node_->get_logger(), "No Path Found !!!!!!!!!!!");
+      raw_path.clear();
+      return false;
+    }
+
+  } else {
+    RCLCPP_INFO(node_->get_logger(), "Straight Line Path Found!");
+    data.push_back({dst.x, dst.y, dist(src.x, src.y, dst.x, dst.y), 0, 0, 1,
+        dist(src.x, src.y, dst.x, dst.y)});
+    curr_id = id_gen;
+  }
+
+  RCLCPP_INFO(node_->get_logger(), "REACHED DEST  %i,   %i", data[curr_id].x, data[curr_id].y);
+
+  backtrace(raw_path, curr_id);
+
+  return true;
+}
+
+void LazyThetaStar::initializePosn()
+{
+  posn.clear();
+  for (int i = 0; i < sizeX * sizeY; i++) {
+    posn.push_back(-1);
+  }
+}
+
+bool LazyThetaStar::losCheck(map_pts & x0, map_pts & y0, map_pts & x1, map_pts & y1) const
+{
+  int dy = abs(y1 - y0), dx = abs(x1 - x0), f = 0;
+  int sx = x1 > x0 ? 1 : -1, sy = y1 > y0 ? 1 : -1;
+  int cx = x0, cy = y0;
+  int ux = (sx - 1) / 2, uy = (sy - 1) / 2;
+
+  if (dx >= dy) {
+    while (cx != x1) {
+      f += dy;
+      if (f >= dx) {
+        if (!isSafe(cx + ux, cy + uy)) {
+          return false;
+        }
+        cy += sy;
+        f -= dx;
+      }
+      if (f != 0 && !isSafe(cx + ux, cy + uy)) {
+        return false;
+      }
+      if (dy == 0 && !isSafe(cx + ux, cy) && !isSafe(cx + ux, cy - 1)) {
+        return false;
+      }
+      cx += sx;
+    }
+  } else {
```

Comment:
Admittedly, I haven't looked through this function in detail to know if its doing the right thing, but I do see `f` being added to `dx` and subtracted from `dy` which seems like that might not be what you want to do. I'm not sure though. You tell me

Reply:
Here f basically acts as an error term, which (for example in the case of dx > dy) is used to track if the number of times that x-coordinate is incremented, is y-coordinate be incremented too, which is tracked by doing, f += dy, and if f becomes greater than dx, then we subtract dx, check, increment y and continue.

Reply:
The increments done for dx >= dy are of the form, 
X: x-coord, Y: y-coord, n: number of iterations to reach f > dx
( X + n*1, Y + (dy * n)/dx )

Reply:
got it - OK. Some comments here would also be good. See other comment below. I think comments here would be greatly helpful for people later coming along to have to modify / improve / change it.